resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -674,6 +671,43 @@ private[spark] class ApplicationMaster(
     allocatorLock.notifyAll()
   }
 
+  private def startAMCredentialRenewer(): Unit = {
+    // If the credentials file config is present, we must periodically renew tokens. So create
+    // a new AMDelegationTokenRenewer
+    if (sparkConf.contains(CREDENTIALS_FILE_PATH.key)) {
+      // If a principal and keytab have been set, use that to create new credentials for executors
+      // periodically
+      val credentialManager = new YARNHadoopDelegationTokenManager(
+        sparkConf,
+        yarnConf,
+        () => YarnSparkHadoopUtil.get.hadoopFSsToAccess(sparkConf, yarnConf))
+
+      val credentialRenewer = new AMCredentialRenewer(sparkConf, yarnConf, credentialManager)
+      credentialRenewer.scheduleLoginFromKeytab()
+    }
+  }
+
+  private def startAMCredentialRenewerThread(): Thread = {
```

Comment:
Can you add comments explaining this is done so that `AMCredentialRenewer` picks up the custom class loader? This thread will be very short-lived, so it would be nice to explain why it's needed.